lib/src/iterable_extensions.dart

```diff
@@ -0,0 +1,519 @@
+  T sum() {
+    var result = (0.5 is T ? 0.0 : 0) as T;
+    for (var value in this) {
+      result += value;
+    }
+    return result;
+  }
+
+  /// The average of the elements.
+  ///
+  /// The average is zero if the iterable is empty.
+  double average() {
+    var result = 0.0;
+    var count = 0;
+    for (var value in this) {
+      result += value;
+      count += 1;
+    }
+    return result / count;
+  }
+}
+
+/// Extensions on iterables of iterables.
+extension IterableIterableExtension<T> on Iterable<Iterable<T>> {
+  /// The the elements of each iterable in this iterable.
+  Iterable<T> get flattened sync* {
+    for (var elements in this) {
+      yield* elements;
+    }
+  }
+}
+
+/// Extensions that apply to iterables of comparable elements.
+extension IterableComparableExtension<T extends Comparable<T>> on Iterable<T> {
+  /// A minimal element of the iterable, or `null` it the iterable is empty.
+  T /*?*/ get minOrNull {
+    var iterator = this.iterator;
+    if (iterator.moveNext()) {
+      var value = iterator.current;
+      while (iterator.moveNext()) {
+        var newValue = iterator.current;
+        if (value.compareTo(newValue) > 0) {
+          value = newValue;
+        }
+      }
+      return value;
+    }
+    return null;
+  }
+
+  /// A minimal element of the iterable.
+  ///
+  /// The iterable must not be empty.
+  T get min {
+    var iterator = this.iterator;
+    if (iterator.moveNext()) {
+      var value = iterator.current;
+      while (iterator.moveNext()) {
+        var newValue = iterator.current;
+        if (value.compareTo(newValue) > 0) {
+          value = newValue;
+        }
+      }
+      return value;
+    }
+    throw StateError("No element");
+  }
+
+  /// A maximal element of the iterable, or `null` if the iterable is empty.
+  T /*?*/ get maxOrNull {
+    var iterator = this.iterator;
+    if (iterator.moveNext()) {
+      var value = iterator.current;
+      while (iterator.moveNext()) {
+        var newValue = iterator.current;
+        if (value.compareTo(newValue) < 0) {
+          value = newValue;
+        }
+      }
+      return value;
+    }
+    return null;
+  }
+
+  /// A maximal element of the iterable.
+  ///
+  /// The iterable must not be empty.
+  T get max {
+    var iterator = this.iterator;
+    if (iterator.moveNext()) {
+      var value = iterator.current;
+      while (iterator.moveNext()) {
+        var newValue = iterator.current;
+        if (value.compareTo(newValue) < 0) {
+          value = newValue;
+        }
+      }
+      return value;
+    }
+    throw StateError("No element");
+  }
+
+  /// Creates a sorted list of the elements of the iterable.
+  ///
+  /// If the [compare] function is not supplied, the sorting uses the
+  /// natural [Comparable] ordering of the elements.
+  List<T> sorted([int compare(T a, T b)]) => [...this]..sort(compare);
+
+    /// Whether the elements are sorted by the [compare] ordering.
```

Comment:
The doc should describe the case where `compare` is omitted.

```dart
  /// Whether the element are sorted.
  ///
  /// The order is checked against [compare] if it is passed,
  /// otherwise it is checked against the element's natural order.
```